petastorm/pytorch.py

```diff
@@ -0,0 +1,118 @@
+#  Copyright (c) 2017-2018 Uber Technologies, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from petastorm.reader import Reader, ShuffleOptions
+from petastorm.workers_pool.thread_pool import ThreadPool
+
+from torch.utils.data.dataloader import default_collate
+
+
+class DataLoader(object):
+    """
+    A data loader adaptor for ``torch.utils.data.DataLoader``.
+
+    This class iterates and returns items from the Reader in batches.
+
+    This loader can be used as a context manager, but it will terminate at the end of an epoch.
+    The context will invoke next_epoch() upon entry.
+
+    If not used as context manager, invoke the next_epoch() function at the start of each epoch, and
+    once more at the very end.
+    """
+
+    def __init__(self, dataset_url, schema_fields=None, batch_size=1, shuffle_options=None,
```

Comment:
This class copies a lot of logic and boilerplate from the Reader class. This will require keeping a lot of things in sync, e.g. arguments, etc. Is there a reason we dont pass in an instantiated reader object to the DataLoader or have it be a parent class?

Reply:
This would also make it more inline with what we do with tensorflow (have it be an adapter on top of the python reader instead of a different class)

Reply:
Hmm, that's a good point.  I also didn't like reproducing reader params all over again.

Also, I did battle these two options out in my mind...didn't think to look at tf for an existing pattern!

OK, let me know reformulate this to pass in an instantiated reader.